src/api/cpp/cvc5.h

```diff
@@ -4155,14 +4265,48 @@ class CVC5_EXPORT Solver
    */
   Statistics getStatistics() const;
 
+  /* .................................................................... */
+  /* Optimization                                                         */
+  /* .................................................................... */
+
+  /**
+   * Adds an optimization objective
+   * @param target the target formula to optimize
+   * @param objType whether it's MAXIMIZE or MINIMIZE
+   * @param bvSigned if the target is of sort bitvector,
+   *  indicate whether it's a signed bitvector or not, defaults to false,
+   *  for other types, it's omitted
+   */
+  void addObjective(Term target, ObjectiveType objType, bool bvSigned = false);
```

Comment:
I'd prefer not adding this method, since it makes it so that objectives are part of the SMT state.

Can we make it so that checkOpt takes a vector of <Term, ObjectiveType> pairs?

Also, I don't think we should have a special flag bvSigned; can we make this part of ObjectiveType?  E.g. have 4 values { Min, Max, SignedMin, SignedMax } ?

Reply:
I understand making it stateful will bring lots of problems in this case -- changing the assertions and other states might affect the optimization states (e.g., asserting new constraints between two runs of Pareto optimization will might get incorrect results), but I think that's inherent though -- if we want Pareto optimization return different valid solutions upon different runs, we need to make something as the state (even if we pass the objectives as arguments, we still need to store the objectives that we have solved previously, so it's also a state). And of course, handling either case would be clumsy anyways (and it's not well-handled yet)... 
As for making optimization objectives as states, it's definitely a design choice -- it doesn't bring us more trouble than handling the Pareto optimization AFAIK (WHEW... luckily) and gives us a possibility to make optimization incremental (same problem as handling Pareto). 
My thought is to make it like assertFormula / checkSat / getValue and addSygusConstraint / checkSynth / getSynthSolution triplets, to make it addObjective / checkOpt / getOptValue (not-yet there...) (and in the future if there's optimal synthesis, might be optSynth or synthOpt / getOptSynthSolution). 
What do you think of it? 

As for the objective types, I totally agree, and will work on it later on.

Reply:
It seems like Pareto + assertFormula in between is the only problem. Now it should handle Pareto correctly.